### PR TITLE
DRAFT ARM32 enable NEON

### PR DESCRIPTION
**Pull request by seanshpark**
on-going draft to enable NEON with ARM32

Signed-off-by: SaeHie Park <saehie.park@gmail.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
For #8379

**Comment by seanshpark** [comment]
@binarman , could you please take a look and share your opinions?

**Comment by seanshpark** [comment]
@binarman , ping

**Comment by seanshpark** [comment]
@binarman , ping :)

**Comment by seanshpark** [comment]
CC @lucenticus @lemmaa , I think @binarman is a bit busy so I hope someone else can give some feedbacks :)

**Comment by lucenticus** [comment]
@AShedko @BalyshevArtem  PTAL

**Review by binarman — COMMENTED** [review]
@seanshpark 
Sorry for delay
Overall looks good to me, but I don't get why you disabled `neon-vfpv4` flag =)
found [this comment](https://github.com/Samsung/ONE/issues/8379#issuecomment-1056052027), but still don't sure what is happening =)

File: compiler/luci-interpreter/pal/linux/pal.cmake [hunk 1]
```diff
@@ -43,6 +43,33 @@ macro(add_pal_to_target TGT)
     set(PAL_SOURCES ${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/tensor_utils.cc
             ${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/reference/portable_tensor_utils.cc
             ${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/quantization_util.cc)
+
+    if(${TARGET_ARM32_NEON_ENABLED})
+        set(PAL_SOURCES ${PAL_SOURCES}
+                ${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/optimized/neon_tensor_utils.cc
+                ${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/optimized/cpu_check.cc)
+        set(PAL_SOURCES ${PAL_SOURCES}
```
[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci-interpreter/pal/linux/pal.cmake:51]: Thanks for the review :)

> why you disabled neon-vfpv4 flag =)

This draft is to enabled `neon-vfpv4` flag.
background: current kernels codes (from tensorflow) don't have NEON source codes.
with `neon-vfpv4` flag, kernel codes will look for NEON codes with NEON compiler directive.
this draft is to provide those NEON codes to fix lots of link errors.

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci-interpreter/pal/linux/pal.cmake:51]: > Why not add all of this in one go?

no particular reason, maybe cause added one by one while fixing link errors. :) I'll put them in one.

**Comment by seanshpark** [comment]
@binarman , I'll start posting PRs, can I add you as a reviewer? or can you recommend someone else?

**Comment by binarman** [comment]
@seanshpark 
I think it is better to add @BalyshevArtem @AShedko , as mentioned @lucenticus .

My review could introduce unwanted delays =/

**Comment by seanshpark** [comment]
Fails with rebase to master

this is without NEON
```
1: FAILED
1: - Conv2D_000
1: - Conv2D_002
1: - Conv2D_003
1: - Conv2D_U8_000
1: - TransposeConv_000
1/2 Test #1: luci_value_cross_test ............***Failed   83.50 sec

2: FAILED
2: - SVDF_000
2: - SVDF_001
2/2 Test #2: luci_value_cross_tol_test ........***Failed    1.46 sec
```

File: compiler/luci-interpreter/src/kernels/L2Pool2D.test.cpp [hunk 2]
```diff
@@ -206,7 +206,8 @@ TEST_F(L2Pool2DTest, FloatPaddingSameStride)
   kernel.execute();
 
   std::vector<float> ref_output_data{3.5, 6.0, 6.5, 5.70088, 2.54951, 7.2111, 8.63134, 7.0};
-  EXPECT_THAT(extractTensorData<float>(output_tensor), FloatArrayNear(ref_output_data));
+  // NOTE with NEON+ruy, error is #1=-1.14441e-05, #6=-1.81198e-05
+  EXPECT_THAT(extractTensorData<float>(output_tensor), FloatArrayNear(ref_output_data, 2.0e-5f));
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/L2Pool2D.test.cpp:210]:
```suggestion
  EXPECT_THAT(extractTensorData<float>(output_tensor), FloatArrayNear(ref_output_data, 1.0e-4f));
```

From TF test code.
``` cpp
  EXPECT_THAT(m.GetOutput(),
              ElementsAreArray(ArrayFloatNear(
                  {3.5, 6.0, 6.5, 5.70088, 2.54951, 7.2111, 8.63134, 7.0},
                  /*max_abs_error=*/1e-4)));
```

**Comment by seanshpark** [comment]
NEON test failure (https://github.com/Samsung/ONE/pull/8578#issuecomment-1080044534) somehow is not reproduced. I think my environment had some problems...

**Comment by seanshpark** [comment]
done

**Pull request closed without merging**